_posts/2019-12-18-package-lock.md

```diff
@@ -0,0 +1,101 @@
+- *I just wrote some code locally, and everything works*
+- *I go ahead and `git push` my changes.  For some reason CI is breaking!?*
+- *It's a lint error... it looks like `prettier` is failing on some formatting change*.
+
+This can happen because **the version of `prettier` you're using locally can be different than the version of `prettier` you're getting in CI**.  This is one of the critical problems package lock files can solve - the version you're getting on your computer, and on CI are (mostly) going to be the same.
+
+Having an exact record of the full transitive dependency tree can be great for a few reasons:
+- Every environment is (mostly) guaranteed to install the same set of dependencies when running `npm install`
+- `npm install` typically runs faster because some resolution steps can be skipped
+- There is a log in source control of the exact tree used to run your code and tests
+
+This all sounds great, but guess what ... *I don't use them, and if you're building libraries, you probably shouldn't use them either*.
+
+## Why lock files are not great
+While lock files _sound_ great, they come with some real baggage ranging from cosmetically annoying, to outright poor software practices.  For context - my team at Google builds npm modules that other developers use to make their applications.  In the context of a library, it does not make sense to use lock files (more on the end application use case later). **We've broken far more customers specifically _because of our lock file usage_, and decided to start ignorning them across the ~80 or so npm modules we ship.**
+
+### Generated files don't belong in git
+The first reason I don't like lock files is largely cosmetic.  *I generally don't like generated files to be tracked in source control*.  I personally don't like this for a few reasons:
+- They create terrible merge conflicts
+- They make code reviews more difficult
+- Things like lock file maintenance PRs muddy commit history
+
+When dealing with PRs and trying to keep them up to date I just find the ergonomics... clunky.
+
+### You're not testing reality
+Way more importantly - **lock files provide false comfort**.  The lock file you check into source control has no material impact on the module that developers will install.  `package-lock.json` is not packaged into your module, and is not delivered to developers when they `npm install` your module.  It is *only* used during local `npm install` for developer machines, and in CI.
+
+In other words - The transitive dependencies that are defined in `package-lock.json` may not at all reflect what developers see when they run `npm install`.  The transitive dependencies you test against in CI may not at all reflect what developers are seeing.  You are probably testing against an entirely different set of dependencies than the ones developers will see.
+
+Can you see how that's a big problem?  Here's how it happens:
```

Comment:
These 3-4 paragraphs seem to be repeatedly drive home the point that lock files are not used by dependors. However, I found it a bit hard to understand this until I spoke to you. How about replacing this section with something a lot more concise, like,

--------------------

### You're not testing what users get

When a library commits their lock file, they're ensuring that the next time they build their library their build will have the exact same transitive dependencies. However, that lock file is not used by _dependors_. That is: dependors of the library may get an entirely different set of dependencies.

So, if your library works with a very specific set of transitive dependencies - and may break with different versions of dependencies (ex diamond dependency problems) - a lock file committed to git _is not used by dependors, and does not stop them running into the same issue_.

-----------------------

WDYT?